docs/OV_Runtime_UG/ov_infer_request.md

```diff
@@ -0,0 +1,255 @@
+
+`ov::InferRequest` can be created from the `ov::CompiledModel`:
+
+@sphinxdirective
+
+.. tab:: C++
+
+    .. doxygensnippet:: docs/snippets/ov_infer_request.cpp
+       :language: cpp
+       :fragment: [create_infer_request]
+
+.. tab:: Python
+
+    .. doxygensnippet:: docs/snippets/ov_infer_request.py
+       :language: python
+       :fragment: [create_infer_request]
+
+@endsphinxdirective
+
+## Run inference
+
+`ov::InferRequest` supports synchronous and asynchronous modes for inference.
+
+### Synchronous mode
+
+You can use `ov::InferRequest::infer()` to infer model in synchronous mode:
+
+@sphinxdirective
+
+.. tab:: C++
+
+    .. doxygensnippet:: docs/snippets/ov_infer_request.cpp
+       :language: cpp
+       :fragment: [sync_infer]
+
+.. tab:: Python
+
+    .. doxygensnippet:: docs/snippets/ov_infer_request.py
+       :language: python
+       :fragment: [sync_infer]
+
+@endsphinxdirective
+
+### Asynchronous mode
+
+Asynchronous mode can improve overall frame-rate of the application, because rather than wait for inference to complete, the app can continue doing things on the host, while accelerator is busy. You can use `ov::InferRequest::start_async()` to infer model in asynchronous mode:
+
+@sphinxdirective
+
+.. tab:: C++
+
+    .. doxygensnippet:: docs/snippets/ov_infer_request.cpp
+       :language: cpp
+       :fragment: [async_infer]
+
+.. tab:: Python
+
+    .. doxygensnippet:: docs/snippets/ov_infer_request.py
+       :language: python
+       :fragment: [async_infer]
+
+@endsphinxdirective
+
+Asynchronous mode supports two ways to wait inference results:
+  * `ov::InferRequest::wait_for()` - specify maximum duration in milliseconds to block for. The method is blocked until the specified timeout has elapsed, or the result becomes available, whichever comes first.
+    @sphinxdirective
+
+    .. tab:: C++
+
+        .. doxygensnippet:: docs/snippets/ov_infer_request.cpp
+           :language: cpp
+           :fragment: [wait_for]
+
+    .. tab:: Python
+
+        .. doxygensnippet:: docs/snippets/ov_infer_request.py
+           :language: python
+           :fragment: [wait_for]
+
+    @endsphinxdirective
+  * `ov::InferRequest::wait()` - waits until inference result becomes available
+    @sphinxdirective
+
+    .. tab:: C++
+
+        .. doxygensnippet:: docs/snippets/ov_infer_request.cpp
+           :language: cpp
+           :fragment: [wait]
+
+    .. tab:: Python
+
+        .. doxygensnippet:: docs/snippets/ov_infer_request.py
+           :language: python
+           :fragment: [wait]
+
+    @endsphinxdirective
+
+Both requests are thread-safe: can be called from different threads without fearing corruption and failures.
+
+Also InferRequest provides an functionality which allows to avoid a call of `ov::InferRequest::wait()`, in order to do it, you can use `ov::InferRequest::set_callback()` method. This method allows to set callback which will be called after completing run of InferRequest.
```

Comment:
let's refer to classification sample async which shows in details how to use callbacks

Reply:
let's also explicitly state that we must not use strong reference of infer_request in callback, otherwise we have cyclic references. We must use weak references like `infer_request*, infer_request&, std::weal_ptr<infer_request>... etc`